### DISCUSSION TITLE
Is mod_auth_openidc able to work with multiple threads sending requests?

### DISCUSSION THREAD

**Discussion by akasha2z** [OpenIDC/mod_auth_openidc#1042] [Q&A]
Hi,

I am using mod_auth_openidc (version 2.14.12.1) with multiple threads sending GET calls using bearer token, using OAuth 2.0.
Occasionally some request fails giving the error "oidc_proto_jwt_verify: JWT signature verification failed: [bedrock/private/x86_64/build/distro/jose.c:995: oidc_jwt_verify]: could not find key with kid: KJiH2qnS1pqB_QB8bt_SImpOMFu-r0eUK_xa70TjkO0"

When I try this with just single thread, it runs fine, no issues.

That is why questioning on multi-thread usage here, is it safe? Can it handle multiple requests at same time?

Note - Refresh interval is set to 300

**Comment by zandbelt** [comment]
mod_auth_openidc is able to handle multiple requests at the same time, in different Apache threads and/or processes; the log entry itself is most likely caused by the Authorization rolling over keys before or after issuing the access token; however that would not explain why it does not happen in a single thread. FWIW: does "a single thread" in your question refer to a single-threaded client, a single-threaded Apache server, or both?

The log entry filename suggests that you've compiled the module yourself but you've not used the standard build scripts, so this may be related to your problem. Why would you need to do this rather than using the stock provided packages from your distro or the Github releases page here?

To analyse this further you'll need to provide the complete set of OIDC config primitives that you're using (what does "refresh interval" refer to?) and the platform/OS/distribution that you're on. Also, it would be best to provide a Dockerfile to reproduce this with the binaries from the releases page.

On a final note: if you're only using OAuth 2.0 RS functionality it is better to use [mod_oauth2](https://github.com/OpenIDC/mod_oauth2) since this functionality is deprecated in mod_auth_openidc.
